client-course-schedulizer/src/utilities/hooks/useImportFile.ts

```diff
@@ -0,0 +1,35 @@
+import { isEqual } from "lodash";
+import { ChangeEvent, useContext } from "react";
+import { AppContext } from "../services/appContext";
+import * as readCSV from "../helpers/readCSV";
+
+// A closure with statefulness. Used to handle changes to inputs
+export const useImportFile = () => {
+  const {
+    appState: { schedule },
+    appDispatch,
+    setIsCSVLoading,
+  } = useContext(AppContext);
+
+  // TODO: this only runs when input changes, but if the same file
+  // is uploaded, this will not run.
+  // https://stackoverflow.com/questions/5201317/read-the-contents-of-a-file-object
+  const onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
+    setIsCSVLoading(true);
+    const file: Blob | null = e.target.files && e.target.files[0];
+    const read = new FileReader();
+    file && read.readAsBinaryString(file);
+
+    read.onloadend = () => {
+      const scheduleJSON = readCSV.csvStringToSchedule(String(read.result));
+      // TODO: store in local storage incase prof navigates away while editing.
+      // currently a redundant check
+      if (!isEqual(schedule, scheduleJSON)) {
+        appDispatch({ payload: { schedule: scheduleJSON }, type: "setScheduleData" });
```

Comment:
Yeah, I'm not sure. Maybe we could just await every appDispatch call just to be safe? I didn't think it would be an issue but it has something to do with how massive the schedule is. 

I don't understand the bevahior. But I can add the await's everywhere and see what happens.